proposals/0071-Entity/06-querying.md

```diff
@@ -0,0 +1,555 @@
+
+**Query:**
+```promql
+container_cpu_usage_seconds_total{k8s.namespace.name="production"}
+```
+
+**Before enrichment (raw series from storage):**
+```
+container_cpu_usage_seconds_total{
+    container="nginx",
+    k8s.namespace.name="production",
+    k8s.pod.uid="abc-123",
+    k8s.node.uid="node-001"
+} 1234.5
+```
+
+**After enrichment (returned to user):**
+```
+container_cpu_usage_seconds_total{
+    # Original metric labels
+    container="nginx",
+    
+    # Identifying labels (correlation keys, already on series)
+    k8s.namespace.name="production",
+    k8s.pod.uid="abc-123",
+    k8s.node.uid="node-001",
+    
+    # Descriptive labels from k8s.pod entity
+    k8s.pod.name="nginx-7b9f5",
+    k8s.pod.status.phase="Running",
+    k8s.pod.start_time="2024-01-15T10:30:00Z",
+    
+    # Descriptive labels from k8s.node entity
+    k8s.node.name="worker-1",
+    k8s.node.os="linux",
+    k8s.node.kernel.version="5.15.0"
+} 1234.5
+```
+
+### Enrichment Algorithm
+
+```go
+func (ev *evaluator) enrichSeries(
+    ctx context.Context,
+    series storage.Series,
+    timestamp int64,
+) labels.Labels {
+    originalLabels := series.Labels()
+    
+    // 1. Find correlated entities via storage index
+    entityRefs := ev.entityQuerier.EntitiesForSeries(series.Ref())
+    
+    if len(entityRefs) == 0 {
+        return originalLabels // No entities, return as-is
+    }
+    
+    // 2. Build enriched label set
+    builder := labels.NewBuilder(originalLabels)
+    
+    for _, entityRef := range entityRefs {
+        entity := ev.entityQuerier.GetEntity(entityRef)
+        
+        // Get descriptive labels at the sample timestamp
+        descriptiveLabels := entity.DescriptiveLabelsAt(timestamp)
+        
+        // Merge into result
+        descriptiveLabels.Range(func(l labels.Label) {
+            builder.Set(l.Name, l.Value)
+        })
+    }
+    
+    return builder.Labels()
+}
+```
+
+---
+
+## Filtering by Entity Labels
+
+Since entity labels appear as labels in query results, standard PromQL label matchers work:
+
+### By Identifying Labels
+
+```promql
+# Filter by pod UID (identifying)
+container_cpu_usage_seconds_total{k8s.pod.uid="abc-123"}
+```
+
+This is efficient because identifying labels are stored on the series and indexed.
+
+### By Descriptive Labels
+
+```promql
+# Filter by pod name (descriptive)
+container_cpu_usage_seconds_total{k8s.pod.name="nginx-7b9f5"}
+
+# Filter by node OS (descriptive)
+container_memory_usage_bytes{k8s.node.os="linux"}
+
+# Regex matching on descriptive labels
+http_requests_total{service.version=~"2\\..*"}
+```
+
+**Query Execution for Descriptive Label Filters:**
+
+1. Select all series that might match (based on metric name and any indexed labels)
+2. For each series, look up correlated entities
+3. Get descriptive labels at evaluation timestamp
+4. Apply the filter: keep series where enriched labels match
```

Comment:
It should be noted that this is not very efficient in this naive form.

I would assume that a classical join query is probably already more efficient now. But in whatever way things are implemented, this can be done more efficiently, at least behind the scenes. Spelling out the least efficient way of doing it as if it were a satisfactory solution is misleading. This should at the very least mention that this can (and most likely) has to be improved.